### PR TITLE
Docs Update device memory pages

### PR DESCRIPTION
**Pull request by MKKnorr**

### CODE REVIEW

File: docs/how-to/hip_runtime_api/memory_management/device_memory/texture_fetching.rst [hunk 1]
```diff
@@ -5,56 +5,67 @@
...
 
   Texture used as example
 
-Texture sampling
-===============================================================================
+In HIP textures objects are of type :cpp:struct:`hipTextureObject_t` and created
+using :cpp:func:`hipCreateTextureObject`.
+
+For a full list of available texture functions see the `HIP texture API
+reference <../../../../doxygen/html/group___texture.html>`_.
```
Comment (neon60) [inline docs/how-to/hip_runtime_api/memory_management/device_memory/texture_fetching.rst]: Please use the :ref:`texture_management_reference`

**Comment by neon60** [comment]
No other comments. LGTM.

**Review by neon60 — APPROVED** [review]

File: docs/how-to/hip_runtime_api/memory_management/device_memory.rst [hunk 2]
```diff
@@ -1,52 +1,280 @@
...
-memory (VRAM), and is accessible by the kernels operating on the device. Recent
-architectures use graphics double data rate (GDDR) synchronous dynamic
-random-access memory (SDRAM) such as GDDR6, or high-bandwidth memory (HBM) such
-as HBM2e. Device memory can be allocated as global memory, constant, texture or
-surface memory.
+Device memory is random access memory that is physically located on a GPU. In
+general it is memory with a bandwidth that is an order of magnitude higher
+compared to RAM available to the host. That high bandwidth is only available to
+on-device accesses, accesses from the host or other devices have to go over a
+special interface, usually the PCIe bus or AMD Infinity Fabric, which is
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/memory_management/device_memory.rst]:
```suggestion
special interface which is considerably slower, usually the PCIe bus or AMD Infinity Fabric.
```

File: docs/how-to/hip_runtime_api/memory_management/device_memory.rst [hunk 3]
```diff
@@ -1,52 +1,280 @@
...
+considerably slower.
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/memory_management/device_memory.rst]:
```suggestion
```

File: docs/how-to/hip_runtime_api/memory_management/device_memory.rst [hunk 4]
```diff
@@ -1,52 +1,280 @@
...
+
+On certain architectures like APUs, the GPU and CPU share the same physical
+memory.
+
+There is also a special local data share on-chip directly accessible to the
+:ref:`compute units <hardware_implementation>`, that can be used for shared
+memory.
+
+The physical device memory can be used to back up several different memory
+spaces in HIP:
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/memory_management/device_memory.rst]:
```suggestion
spaces in HIP, as described below. 
```

File: docs/how-to/hip_runtime_api/memory_management/device_memory.rst [hunk 5]
```diff
@@ -1,52 +1,280 @@
...
+The specified amount of memory is allocated by each thread that executes the
+instructions. The recommended way to allocate the memory depends on the use
+case. If the memory is intended to be shared between the threads of a block, it
+is generally beneficial to allocate on large block of memory, due the way the
+memory is accessed.
+
+.. note::
+   Memory allocated within a kernel can only be freed in kernels, not by the HIP
+   runtime on the host, like :cpp:func:`hipFree`. It is also not possible to
+   free device memory allocated on the host, with e.g. :cpp:func:`hipMalloc`, in
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/memory_management/device_memory.rst]:
```suggestion
   free device memory allocated on the host, with :cpp:func:`hipMalloc` for example, in
```

File: docs/how-to/hip_runtime_api/memory_management/device_memory.rst [hunk 6]
```diff
@@ -1,52 +1,280 @@
...
+be specified. They are represented by ``hipTextureObject_t`` and created using
+:cpp:func:`hipCreateTextureObject`.
+
+The underlying memory is a 1D, 2D or 3D ``hipArray_t``, that needs to be
+allocated using :cpp:func:`hipMallocArray`.
+
+On the device side, texture objects are accessed using the ``tex1D/2D/3D``
+functions.
+
+The texture management functions can be found in the :ref:`texture management
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/memory_management/device_memory.rst]:
```suggestion
The texture management functions can be found in the :ref:`Texture management
```

File: docs/how-to/hip_runtime_api/memory_management/device_memory.rst [hunk 7]
```diff
@@ -1,52 +1,280 @@
...
-A read-write version of texture memory, which can be useful for applications
-that require direct manipulation of 1D, 2D, or 3D hipArray_t. 
+A read-write version of texture memory. It is created in the same way as a
+texture, but with :cpp:func:`hipCreateSurfaceObject`.
+
+Since surfaces also cached in the read-only texture cache, the changes written
+back to the surface can't be observed in the same kernel. A new kernel has to be
+launched in order to see the updated surface.
+
+The corresponding functions are listed in the :ref:`surface object API reference
```
Comment (randyh62) [inline docs/how-to/hip_runtime_api/memory_management/device_memory.rst]:
```suggestion
The corresponding functions are listed in the :ref:`Surface object API reference
```

**Review by randyh62 — APPROVED** [review]
Left a few comments. Looks good overall.

File: docs/how-to/hip_runtime_api/memory_management/device_memory.rst [hunk 8]
```diff
@@ -1,52 +1,280 @@
...
+
+It can be allocated from the host via the :ref:`HIP runtime memory management
+functions <memory_management_reference>` like :cpp:func:`hipMalloc`, or can be
+defined using the ``__device__`` qualifier on variables.
+
+It can also be allocated within a kernel using ``malloc`` or ``new``.
+The specified amount of memory is allocated by each thread that executes the
+instructions. The recommended way to allocate the memory depends on the use
+case. If the memory is intended to be shared between the threads of a block, it
+is generally beneficial to allocate on large block of memory, due the way the
```
Comment (TomSang) [inline docs/how-to/hip_runtime_api/memory_management/device_memory.rst]: due to the way...

File: docs/how-to/hip_runtime_api/memory_management/device_memory.rst [hunk 9]
```diff
@@ -1,52 +1,280 @@
...
+  __syncthreads();
+
+  // load pointer into thread-local variable to avoid
+  // unnecessary accesses to shared memory
+  int *localPtr = memory;
+
+  // work with allocated memory, e.g. initialization
+  for(int i = 0; i < elementsPerThread; ++i){
+    // access in a contiguous way
+    localPtr[i * blockSize + threadIdx.x] = i;
```
Comment (tomsang) [inline docs/how-to/hip_runtime_api/memory_management/device_memory.rst:91]: Not sure of the purpose of this logics.
Anyway,  kernel will crash if blockSize -threadIdx.x <1

File: docs/how-to/hip_runtime_api/memory_management/device_memory.rst [hunk 10]
```diff
@@ -1,52 +1,280 @@
...
+kernel.
+
+.. code-block:: cpp
+
+  extern __shared__ int dynamic_shared[];
+  __global__ void kernel(int sizeX, int sizeY, int totalSize){
+    int* array1 = dynamic_shared;
+    // array1 is interpreted as 2D of size:
+    int array1_size = sizeX * sizeY;
+    int* array2 = &(array[sizeX * sizeY]);
```
Comment (TomSang) [inline docs/how-to/hip_runtime_api/memory_management/device_memory.rst]: Shoud it be the following?
int* array2 = &(array1[sizeX * sizeY]);

**Review by TomSang — CHANGES_REQUESTED** [review]

File: docs/how-to/hip_runtime_api/memory_management/device_memory.rst (hunk 8 above)
Comment (MKKnorr) [reply docs/how-to/hip_runtime_api/memory_management/device_memory.rst]: Fixed

File: docs/how-to/hip_runtime_api/memory_management/device_memory.rst (hunk 10 above)
Comment (MKKnorr) [reply docs/how-to/hip_runtime_api/memory_management/device_memory.rst]: Yep, thanks. I updated the whole snippet to be a bit more expressive. Also ran it through the compiler this time to verify it works :)

**Comment by TomSang** [comment]
LGTM!

**Pull request merged**